airflow/jobs/job.py

```diff
@@ -190,11 +182,10 @@ def heartbeat(
 
         try:
             # This will cause it to load from the db
-            session.merge(self)
+            self._merge_from(Job._fetch_from_db(self, session))
```

Comment:
I tihnk we need to meet for an interactive brainstorming session and discuss what's going on here. I believe there were some non-obvious behaviours of heartbeat that we should replicate (especially around edge cases).

I tried recently to understand what's going on in this method and explain it here:  https://github.com/apache/airflow/discussions/33689#discussioncomment-6820251

And I think it's not entirely obvious if we have the same now.

Let's discuss on slack @mhenc.

Reply:
@potiuk any update on this?